src/graph/heterograph.cc

```diff
@@ -260,6 +261,155 @@ HeteroGraphPtr HeteroGraph::CopyTo(HeteroGraphPtr g, const DLContext& ctx) {
                                         hgindex->num_verts_per_type_));
 }
 
+std::string HeteroGraph::SharedMemName() const {
+  return shared_mem_ ? shared_mem_->GetName() : "";
+}
+
+HeteroGraphPtr HeteroGraph::CopyToSharedMem(
+      HeteroGraphPtr g, const std::string& name, const std::vector<std::string>& ntypes,
+      const std::vector<std::string>& etypes, const std::set<std::string>& fmts) {
+  auto hg = std::dynamic_pointer_cast<HeteroGraph>(g);
+  CHECK_NOTNULL(hg);
+  if (hg->SharedMemName() == name)
+    return g;
+
+  // Copy buffer to share memory
+  auto mem = std::make_shared<SharedMemory>(name);
+  auto mem_buf = mem->CreateNew(SHARED_MEM_METAINFO_SIZE_MAX);
+  dmlc::MemoryFixedSizeStream ofs(mem_buf, SHARED_MEM_METAINFO_SIZE_MAX);
+  dmlc::SeekStream *strm = &ofs;
+
+  bool has_coo = fmts.find("coo") != fmts.end();
+  bool has_csr = fmts.find("csr") != fmts.end();
+  bool has_csc = fmts.find("csc") != fmts.end();
+  strm->Write(g->NumBits());
+  strm->Write(has_coo);
+  strm->Write(has_csr);
+  strm->Write(has_csc);
+  strm->Write(ImmutableGraph::ToImmutable(hg->meta_graph_));
+  strm->Write(hg->num_verts_per_type_);
+
+  std::vector<HeteroGraphPtr> relgraphs(g->NumEdgeTypes());
+
+  for (dgl_type_t etype = 0 ; etype < g->NumEdgeTypes() ; ++etype) {
+    strm->Write(hg->NumEdges(etype));
+    aten::COOMatrix coo;
+    aten::CSRMatrix csr, csc;
+    std::string prefix = name + "_" + std::to_string(etype);
+    if (has_coo) {
+      coo = hg->GetCOOMatrix(etype).CopyToSharedMem(prefix + "_coo");
+      strm->Write(coo.row_sorted);
+      strm->Write(coo.col_sorted);
+    }
+    if (has_csr) {
+      csr = hg->GetCSRMatrix(etype).CopyToSharedMem(prefix + "_csr");
+      strm->Write(csr.sorted);
+    }
+    if (has_csc) {
+      csc = hg->GetCSCMatrix(etype).CopyToSharedMem(prefix + "_csc");
+      strm->Write(csc.sorted);
+    }
+    relgraphs[etype] = UnitGraph::CreateHomographFrom(csc, csr, coo, has_csc, has_csr, has_coo);
+  }
+
+  auto ret = std::shared_ptr<HeteroGraph>(
+      new HeteroGraph(hg->meta_graph_, relgraphs, hg->num_verts_per_type_));
+  ret->shared_mem_ = mem;
+
+  strm->Write(ntypes);
+  strm->Write(etypes);
+  return ret;
+}
+
+aten::COOMatrix CreateCOOFromSharedMem(const std::string &name, uint8_t nbits,
+    int64_t num_src, int64_t num_dst, int64_t num_edges, bool rsorted, bool csorted) {
+  DLDataType dtype = {kDLInt, nbits, 1};
+  DLContext ctx = {kDLCPU, 0};
+  NDArray row = NDArray::EmptyShared(name+"_row", {num_edges}, dtype, ctx, false);
+  NDArray col = NDArray::EmptyShared(name+"_col", {num_edges}, dtype, ctx, false);
+  return aten::COOMatrix(num_src, num_dst, row, col, aten::NullArray(), rsorted, csorted);
```

Comment:
Where's data ndarray in COO?

Reply:
can we assume coo matrices used in heterograph don't have data array (and csr always have data array)? it simplifies the process.

Reply:
I don't think you can make such assumption.